library/src/main/java/com/datatorrent/lib/async/AbstractAsyncProcessor.java

```diff
@@ -0,0 +1,344 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.datatorrent.lib.async;
+
+import java.util.Iterator;
+import java.util.Queue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import javax.validation.constraints.Min;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Lists;
+
+import com.datatorrent.api.Context;
+import com.datatorrent.api.Operator.IdleTimeHandler;
+import com.datatorrent.common.util.BaseOperator;
+
+
+/**
+ * This base operator can be used to enqueue tuples for asynchronous and parallel processing by separate worker threads.
+ *
+ * User can configure number of threads required for async processing. By default
+ * threads will be created on need basis with no upper limit. These worker threads, spawned by operator thread,
+ * can also be configured to processing timeout.
+ *
+ * <b>Workflow is as follows:</b>
+ * <ol>
+ *   <li>Implementing class can enqueue the tuples of type QUEUETUPLE using <i>enqueueTupleForProcessing</i>
+ *       method.</li>
+ *   <li>When a tuple gets its turn for execution, <i>processTupleAsync</i> gets called
+ *       by worker thread. This is asynchronous call</li>
+ *   <li>Based on maintainTupleOrder parameter, the results are consolidated at the end of every window and respective
+ *       call to <i>handleProcessedTuple</i> will be made. This callback is in operator thread.</li>
+ * </ol>
+ *
+ *
+ * <b>Use case examples:</b>
+ * <ul>
+ *   <li>Parallel reads from external datastore/database system.</li>
+ *   <li>Operations which are long running tasks per tuple but DAG i/o should not be blocked.</li>
+ * </ul>
+ *
+ * @param <QUEUETUPLE> input type of tuple
+ * @param <RESULTTUPLE> Result of async processing of tuple.
+ */
+public abstract class AbstractAsyncProcessor<QUEUETUPLE, RESULTTUPLE> extends BaseOperator implements IdleTimeHandler
+{
+  private static final Logger logger = LoggerFactory.getLogger(AbstractAsyncProcessor.class);
+
+  /**
+   * This is an executor service which holds thread pool and enqueues task for processing.
+   */
+  private transient ExecutorService executor;
+
+  /**
+   * Number of worker threads for processing
+   * 0   - Threads will be created as per need.
```

Comment:
So here is what I've done:
Min = 1 & Max = 32.
32 is the cap on number of threads enforced by javax Max annotation.

Default is set to 32.

Let me know if some other cap on number of threads need to be there.